src/utils/dom-fingerprint.js

```diff
@@ -0,0 +1,120 @@
+ * The order is from the inner to outer nodes in DOM hierarchy.
+ *
+ * If a DOM hierarchy is the following:
+ *
+ * <div id='id1' ...>
+ *   <div id='id2' ...>
+ *     <table ...>       // table:0
+ *       <tr>            // tr:0
+ *         <td>...</td>  // td:0
+ *         <td>          // td:1
+ *           <amp-ad ...></amp-ad>
+ *         </td>
+ *       </tr>
+ *       <tr>...</tr>    // tr:1
+ *     </table>
+ *   </div>
+ * </div>
+ *
+ * With the amp-ad element passed in:
+ * 'amp-ad.0,td.1,tr.0,table.0,div/id2.0,div/id1.0'
+ *
+ * Note: 25 is chosen arbitrarily.
+ *
+ * @param {?Element} element DOM node from which to get fingerprint.
+ * @return {string} Concatenated element ids.
+ */
+export function domFingerprintPlain(element) {
+  const ids = [];
+  let level = 0;
+  while (element && element.nodeType == /* element */ 1 && level < 25) {
+    const id = element.id &&
+               // Skip AMP generated ids.
+               !(element.getResourceId &&
+                 element.id == 'AMP_' + element.getResourceId()) ?
+      `/${element.id}` : '';
+    const nodeName = String(element.nodeName).toLowerCase();
+    ids.push(`${nodeName}${id}${indexWithinParent(element)}`);
+    level++;
+    element = element.parentElement;
+  }
+  return ids.join();
+};
+
+/**
+ * Calculates ad slot DOM fingerprint.  This key is intended to
+ * identify "same" ad unit across many page views. This is
+ * based on where the ad appears within the page's DOM structure.
+ *
+ * @param {?Element} element The DOM element from which to collect
+ *     the DOM chain element IDs.  If null, DOM chain element IDs are not
+ *     included in the hash.
+ * @return {string} The ad unit hash key string.
+ */
+export function domFingerprint(element) {
+  return String(stringHash32(domFingerprintPlain(element)));
+};
+
+/**
+ * Hash function djb2a
+ * This is intended to be a simple, fast hashing function using minimal code.
+ * It does *not* have good cryptographic properties.
+ * @param {string} str
+ * @return {number} 32-bit unsigned hash of the string
+ */
+export function stringHash32(str) {
+  const length = str.length;
+  let hash = 5381;
+  for (let i = 0; i < length; i++) {
+    hash = hash * 33 ^ str.charCodeAt(i);
+  }
+  // Convert from 32-bit signed to unsigned.
+  return hash >>> 0;
+};
+
+
+/**
+ * Gets a string showing the index of an element within
+ * the children of its parent, counting only nodes with the same tag.
+ * Stop at 25, just to have a limit.
+ * @param {!Element} element DOM node to get index of.
+ * @return {string} '.<index>' or ''.
+ */
+function indexWithinParent(element) {
+  // Note that some older browsers return an object for nodeName,
+  // so explicitly coerce to String.
+  const nodeName = String(element.nodeName);
+  // Find my index within my parent's children
+  let count = 0;
+  let sibling = element.previousElementSibling;
+  while (sibling && count < 25) {
```

Comment:
I think it's more important to have an `i` limit (not `count`). Imagine we have a `<span>` preceeded by 10,000 `<div>`s. We'd iterate all of them. But if we had 25 preceding `<span>`s we'd stop? That seems wrong.